packages/flutter/lib/src/widgets/router.dart

```diff
@@ -632,7 +636,10 @@ class _RouterState<T> extends State<Router<T>> with RestorationMixin {
       .then<T>(_verifyRouteInformationParserStillCurrent(_currentRouteInformationParserTransaction, widget))
       .then<void>(delegateRouteSetter())
       .then<void>(_verifyRouterDelegatePushStillCurrent(_currentRouterDelegateTransaction, widget))
-      .then<void>(_rebuild);
+      .then<void>(_rebuild)
+      .onError<_DiscardTransaction>((_DiscardTransaction e, StackTrace trace) {
```

Comment:
cc @Hixie who wrote the code i copied from. I am not sure whether  this may have performance implication, but this is the only way i can think of to skip the`then` future.

Reply:
Could we refactor this into something like 

```dart
widget.routeInformationParser!
      .parseRouteInformation(information).then(_processParsedRouteInfo);
```

with:

```dart
Future<void> _processParsedRouteInfo(T data) async {
   if (route information parser no longer current) {
     return;
   }
   await _processParsedRouteInfo();
   if (route delegate push still current) {
     _rebuild();
   }
}
```

Doing it with exceptions feels a bit strange...